fedimint-core/src/encoding/mod.rs

```diff
@@ -248,17 +249,99 @@ macro_rules! impl_encode_decode_tuple {
     );
 }
 
+/// Specialized version of Encodable for bytes
+pub fn consensus_encode_bytes<W: std::io::Write>(
+    bytes: &[u8],
+    writer: &mut W,
+) -> Result<usize, Error> {
+    let mut len = 0;
+    len += (bytes.len() as u64).consensus_encode(writer)?;
+    writer.write_all(bytes)?;
+    len += bytes.len();
+    Ok(len)
+}
+
+/// Specialized version of Encodable for static byte arrays
+pub fn consensus_encode_bytes_static<const N: usize, W: std::io::Write>(
+    bytes: &[u8; N],
+    writer: &mut W,
+) -> Result<usize, Error> {
+    writer.write_all(bytes)?;
+    Ok(bytes.len())
+}
+
+struct ReadBytesFromFiniteReaderOpts {
+    len: usize,
+    chunk_size: usize,
+}
+
+/// Read `opts.len` bytes from reader, where `opts.len` could potentially be
+/// malicious.
+///
+/// Adapted from <https://github.com/rust-bitcoin/rust-bitcoin/blob/e2b9555070d9357fb552e56085fb6fb3f0274560/bitcoin/src/consensus/encode.rs#L659>
+#[inline]
+fn read_bytes_from_finite_reader<D: Read + ?Sized>(
+    d: &mut D,
+    mut opts: ReadBytesFromFiniteReaderOpts,
+) -> Result<Vec<u8>, io::Error> {
+    let mut ret = vec![];
+
+    assert_ne!(opts.chunk_size, 0);
+
+    while opts.len > 0 {
+        let chunk_start = ret.len();
+        let chunk_size = core::cmp::min(opts.len, opts.chunk_size);
+        let chunk_end = chunk_start + chunk_size;
+        ret.resize(chunk_end, 0u8);
+        d.read_exact(&mut ret[chunk_start..chunk_end])?;
+        opts.len -= chunk_size;
+    }
+
+    Ok(ret)
+}
+
+/// Specialized version of Decodable for bytes
+pub fn consensus_decode_bytes<D: std::io::Read>(r: &mut D) -> Result<Vec<u8>, DecodeError> {
+    let len = u64::consensus_decode(r, &Default::default())?;
+
+    let len: usize =
+        usize::try_from(len).map_err(|_| DecodeError::from_str("size exceeds memory"))?;
+
+    let opts = ReadBytesFromFiniteReaderOpts {
+        len,
+        chunk_size: 64 * 1024,
+    };
+
+    read_bytes_from_finite_reader(r, opts).map_err(DecodeError::from_err)
+}
+
+/// Specialized version of Decodable for fixed-size byte arrays
+pub fn consensus_decode_bytes_static<const N: usize, D: std::io::Read>(
+    r: &mut D,
+) -> Result<[u8; N], DecodeError> {
+    let mut bytes = [0u8; N];
+    r.read_exact(bytes.as_mut_slice())
+        .map_err(DecodeError::from_err)?;
+    Ok(bytes)
+}
+
 impl_encode_decode_tuple!(T1, T2);
 impl_encode_decode_tuple!(T1, T2, T3);
 impl_encode_decode_tuple!(T1, T2, T3, T4);
 
 impl<T> Encodable for &[T]
```

Comment:
we should impl for `[T]` instead of `&[T]`

Reply:
but using Any still doesn't work
because you can't do .downcast_ref::<[u8]>(), because downcast_ref needs `Sized`

Reply:
we could impl for Vec<u8> without unsafe, but not for `Box<[u8]>` or Arc, Rc